hlc: use PTP user space API as HLC clock

darinpp opened the pull request:
Adds support for Linux for using PTP user space API clock
device for HLC. This is needed in case that the host clock is
prone to large jumps (as is the case when using vMotion).

Release note (cli change): added new start option --clock-device that
allows HLC to use PTP user space API clock

knz commented on pkg/base/config.go:
```diff
@@ -213,6 +213,9 @@ type Config struct {
 	RPCHeartbeatInterval time.Duration
+
+	// Enables the use of an IEEE 1588 PTP clock device for HLC current time
+	ClockDevice        string
```
This needs to be reset in `InitDefaults()` below.

darinpp replied:
done

knz commented on pkg/cli/flags.go:
```diff
@@ -51,6 +51,7 @@ var serverSQLAddr, serverSQLPort string
 var serverSQLAdvertiseAddr, serverSQLAdvertisePort string
 var serverHTTPAddr, serverHTTPPort string
 var localityAdvertiseHosts localityList
+var clockDevice string
```
This needs to be reset to its default value in `context.go`. However see my next comment.

darinpp replied:
removed

knz commented on pkg/cli/flags.go:
```diff
@@ -358,6 +359,7 @@ func init() {
 		VarFlag(f, &serverCfg.Stores, cliflags.Store)
 		VarFlag(f, &serverCfg.StorageEngine, cliflags.StorageEngine)
 		VarFlag(f, &serverCfg.MaxOffset, cliflags.MaxOffset)
+		StringFlag(f, &clockDevice, cliflags.ClockDevice, "")
```
Was there a reason not to do `StringFlag(f, &serverCfg.ClockDevice, ...)` and avoid the separate global var?

darinpp replied:
I changed it as you are suggesting.

knz commented on build/builder/mkrelease.sh:
```diff
@@ -39,7 +39,7 @@ case "${1-}" in
       XCMAKE_SYSTEM_NAME=Linux
       TARGET_TRIPLE=x86_64-unknown-linux-gnu
-      LDFLAGS="-static-libgcc -static-libstdc++"
+      LDFLAGS="-static-libgcc -static-libstdc++ -lrt"
```
Does your patch work for the other targets supported below? What does the `build/builder.sh mkrelease` command says for the various targets?

darinpp replied:
It does work for `darwin`, `windows` and `linux-musl` too. 
`linux-gnueabi` and `linux-msan` don't work for me for the unmodified master branch.

knz reviewed (commented):
Only commenting on the CLI/build aspects of this. I'll let andrei or the other team members comment on the HLC changes.

andreimatei reviewed (commented):
nit: are you sure you want to mention IEEE 1588 in the commit title and release notes? I think it's misleading. We're technically using the Linux "PTP hardware clock user space API" (as per here https://www.kernel.org/doc/Documentation/ptp/ptp.txt), but the time source doesn't need to have anything to do with the Precision Time Protocol. I'd just make reference to this API, and otherwise downplay the PTP part.

nit: please make the release note more of a full sentence, and mention the new flag. See this for some guidelines: https://wiki.crdb.io/wiki/spaces/CRDB/pages/186548364/Release+notes
Also, the `Release note()` wants to have one of a predefined categories. The "low risk new functionality" part belongs in the `Release justification` stanza.


pls see comment about the "global state"

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-:-M3x-Dk19ZsAmYET9Jdf:b-89w29j)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @darinpp)

___
*[build/builder/mkrelease.sh, line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3xCDQZ3Z31rFXgZGQP:-M3xCDQZ3Z31rFXgZGQQ:b-o5eb8x) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/build/builder/mkrelease.sh#L42)):*
> ```Shell
>       XCMAKE_SYSTEM_NAME=Linux
>       TARGET_TRIPLE=x86_64-unknown-linux-gnu
>       LDFLAGS="-static-libgcc -static-libstdc++ -lrt"
> ```

pls add a comment somewhere about why linking this is necessary? Do we really not get all of `time.h` without it?
___
*[pkg/base/config.go, line 217 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x0gKWDOPc5_i9dsT-:-M3x0gKXCC1VwBa7auvX:briwurr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/base/config.go#L217)):*
> ```Go
> 	RPCHeartbeatInterval time.Duration
> 
> 	// Enables the use of an IEEE 1588 PTP clock device for HLC current time
> ```

nit: missing period. 
___
*[pkg/base/config.go, line 218 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x0tBUENaAGFga8T3_:-M3x0tBUENaAGFga8T3a:b-yo1kpc) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/base/config.go#L218)):*
> ```Go
> 
> 	// Enables the use of an IEEE 1588 PTP clock device for HLC current time
> 	ClockDevice        string
> ```

please put `Path` in the name, and otherwise comment that this is a path. Also please make reference to the API we're using. Something to include the `fd<<3 | 3` convention. In the worst case, a commit in the linux kernel that introduced it. I found some stuff on it in the past with some googling.
___
*[pkg/cli/cliflags/flags.go, line 679 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x1Trh7PQ22NbwbBmn:-M3x1Trh7PQ22NbwbBmo:b-96wqu4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/cli/cliflags/flags.go#L679)):*
> ```Go
> 		Name: "clock-device",
> 		Description: `
> Use IEEE 1588 PTP clock device for HLC current time. 
> ```

I'd get rid of IEEE 1588 and make reference to some description of the API the device needs to implement.
Also please put more words in the description. Say something about this overriding the "regular" clock.
___
*[pkg/cli/cliflags/flags.go, line 680 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3xCa4w1mOcpRD_yfpy:-M3xCa4w1mOcpRD_yfpz:bc0yx4c) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/cli/cliflags/flags.go#L680)):*
> ```Go
> 		Description: `
> Use IEEE 1588 PTP clock device for HLC current time. 
> <PRE>
> ```

Please say that this is only available on Linux (right?). 
Also consider adding some validation to error out when the flag is used on other platforms (besides the validation based on the version of `UseClockDevice` that gets compiled in - where `+build !windows` is not strictly restricted to linux).
___
*[pkg/server/status/runtime.go, line 507 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x5rHy0NnW_qgtwSe3:-M3x5rHy0NnW_qgtwSe4:b4hw8pj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/server/status/runtime.go#L507)):*
> ```Go
> 	log.Infof(ctx, "runtime stats: %s RSS, %d goroutines, %s/%s/%s GO alloc/idle/total%s, "+
> 		"%s/%s CGO alloc/total, %.1f CGO/sec, %.1f/%.1f %%(u/s)time, %.1f %%gc (%dx), "+
> 		"%s/%s (r/w)net, timeutil.Now calls (all: %d, hlc: %d)",
> ```

these lines are already hard to read; you sure you want this new information? I think it's useful only when developing. I'd put it in a separate `VEventf(2)` message.

Or, rather, I'd get rid of the counters. I don't think it's worth the code. And I'd rather not worry about the performance impact of the atomic increment either. For such a common increment we might want to look at per-cpu counters (which can be hacked in go with `sync.Pool` I think).
___
*[pkg/util/hlc/hlc.go, line 122 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x8DZGEA1vpbOuTJMq:-M3x8DZHA5r49BZch-7-:b-rz4p1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/util/hlc/hlc.go#L122)):*
> ```Go
> // c := hlc.NewClock(hlc.UnixNano, ...).
> func UnixNano() int64 {
> 	return timeutil.HlcNow().UnixNano()
> ```

nit: s/Hlc/HLC everywhere
___
*[pkg/util/timeutil/now_unix.go, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x5SUo0IK-MMBEtJtL:-M3x5SUo0IK-MMBEtJtM:bi7o5ch) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/util/timeutil/now_unix.go#L28)):*
> ```Go
> 	"time"
> 
> 	errors2 "github.com/pkg/errors"
> ```

just one `errors` pkg pls. The right one to use is `cockroachdb/errors`
___
*[pkg/util/timeutil/now_unix.go, line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x26lPEWlLnkqIgiAV:-M3x26lPEWlLnkqIgiAW:b-rxsja) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/util/timeutil/now_unix.go#L33)):*
> ```Go
> var (
> 	clockDevice   *os.File
> 	clockDeviceId uintptr
> ```

as the linter says, make it `clockDeviceID`
___
*[pkg/util/timeutil/now_unix.go, line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x2IM_7jRfLVb-Lvl7:-M3x2IM_7jRfLVb-Lvl8:b-pe3r48) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/util/timeutil/now_unix.go#L47)):*
> ```Go
> 
> // Use the given device as an HLC clock source.
> func UseClockDevice(clockDeviceName string) error {
> ```

Name? Path...
___
*[pkg/util/timeutil/now_unix.go, line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x9Zbc7biYFjagOMJb:-M3x9Zbc7biYFjagOMJc:b6uargr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/util/timeutil/now_unix.go#L47)):*
> ```Go
> 
> // Use the given device as an HLC clock source.
> func UseClockDevice(clockDeviceName string) error {
> ```

I don't like how this clock device works because it changes global state. And we really don't like globals. They always cause troubles - for example you can't run two servers in a process easily any more.
Everything should be scoped to a server. I think the clock device should be used to initialize an hlc clock - either make `hlc.NewClock()` take the device directly, or pass in a closure that uses the respective device as the existing `physicalClock` argument.

Also, I'd try to keep the new per-OS code out of `now.go`, in favor of keeping it in the hlc package. In fact, it seems to me like now.go can be un-split; there no longer seems to be a need for `now_windows.go` according to the comments in it. I think this would be a great time to see if we can clean that up.
___
*[pkg/util/timeutil/now_unix.go, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x6MxRBTdy1WTQvC5A:-M3x6MxRBTdy1WTQvC5B:b-xfrp0n) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/util/timeutil/now_unix.go#L54)):*
> ```Go
> 	clockDevice, err = os.Open(clockDeviceName)
> 	if err == nil {
> 		clockDeviceFd := clockDevice.Fd()
> ```

`clockDeviceFD`
___
*[pkg/util/timeutil/now_unix.go, line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x2N8fCmIOaAkaa-s4:-M3x2N8g1FIibSnzu1gm:b-p2k01a) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/util/timeutil/now_unix.go#L55)):*
> ```Go
> 	if err == nil {
> 		clockDeviceFd := clockDevice.Fd()
> 		clockId := (^clockDeviceFd << 3) | 3
> ```

`clockID`
___
*[pkg/util/timeutil/now_unix.go, line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x55Pq9haDT221-5SY:-M3x55Pq9haDT221-5SZ:bvm5eci) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/util/timeutil/now_unix.go#L55)):*
> ```Go
> 	if err == nil {
> 		clockDeviceFd := clockDevice.Fd()
> 		clockId := (^clockDeviceFd << 3) | 3
> ```

please include a link explaining this magic here too
___
*[pkg/util/timeutil/now_unix.go, line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x6BruAG1BkCTSk8ey:-M3x6BruAG1BkCTSk8ez:bwng1tb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/util/timeutil/now_unix.go#L68)):*
> ```Go
> 		clockDeviceId = clockId
> 	} else {
> 		return fmt.Errorf("Can't open %s, err %s\n", clockDeviceName, err)
> ```

`errors.Errorf()` pls. And don't capitalize the message.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @darinpp)

___
*[pkg/base/config.go, line 218 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x0tBUENaAGFga8T3_:-M3xObF_Ddkp62elh23h:bcnw4n8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/base/config.go#L218)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please put `Path` in the name, and otherwise comment that this is a path. Also please make reference to the API we're using. Something to include the `fd<<3 | 3` convention. In the worst case, a commit in the linux kernel that introduced it. I found some stuff on it in the past with some googling.
</blockquote></details>

This is what I had found. Don't know what happened to man page the patch touches, as it doesn't seem to have this content.
https://lore.kernel.org/patchwork/patch/868609/


<!-- Sent from Reviewable.io -->

darinpp committed:
hlc: use PTP user space API as HLC clock

Adds support for Linux for using PTP user space API clock
device for HLC. This is needed in case that the host clock is
prone to large jumps (as is the case when using vMotion).

Release note (cli change): added new start option --clock-device that
alows HLC to use PTP user space API clock

darinpp reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @knz)

___
*[build/builder/mkrelease.sh, line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3xCDQZ3Z31rFXgZGQP:-M4CyNS-8m_YCN9iYsdY:b6s6r6w) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/build/builder/mkrelease.sh#L42)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

pls add a comment somewhere about why linking this is necessary? Do we really not get all of `time.h` without it?
</blockquote></details>

It seems like that with the library versions that we are using it has to be explicit
___
*[pkg/base/config.go, line 217 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x0gKWDOPc5_i9dsT-:-M4Cz9MyEs2XfNwk_AEI:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/base/config.go#L217)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: missing period. 
</blockquote></details>

fixed
___
*[pkg/base/config.go, line 218 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x0tBUENaAGFga8T3_:-M4CzFow-1-6y9uLCRok:b-e0fh1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/base/config.go#L218)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

This is what I had found. Don't know what happened to man page the patch touches, as it doesn't seem to have this content.
https://lore.kernel.org/patchwork/patch/868609/
</blockquote></details>

I added the path and also the reference to the API in ClockDevice
___
*[pkg/cli/cliflags/flags.go, line 679 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x1Trh7PQ22NbwbBmn:-M4CzQvm5MK1HblMPbPP:b-jk2s9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/cli/cliflags/flags.go#L679)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'd get rid of IEEE 1588 and make reference to some description of the API the device needs to implement.
Also please put more words in the description. Say something about this overriding the "regular" clock.
</blockquote></details>

removed  IEEE 1588
___
*[pkg/cli/cliflags/flags.go, line 680 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3xCa4w1mOcpRD_yfpy:-M4CzU8I2tpXYQoyMmRO:b5ubaww) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/cli/cliflags/flags.go#L680)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Please say that this is only available on Linux (right?). 
Also consider adding some validation to error out when the flag is used on other platforms (besides the validation based on the version of `UseClockDevice` that gets compiled in - where `+build !windows` is not strictly restricted to linux).
</blockquote></details>

Also mentioned that it is only for Linux. Also fixed the `+build` to be `linux` vs `!linux`. The validation is done already in `clock_gettime`. If you are using an incorrect device or you don't have such driver - it gives you back an error. 
___
*[pkg/server/status/runtime.go, line 507 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x5rHy0NnW_qgtwSe3:-M4D-1uT9RI27lgiKs-e:b1iuxee) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/server/status/runtime.go#L507)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

these lines are already hard to read; you sure you want this new information? I think it's useful only when developing. I'd put it in a separate `VEventf(2)` message.

Or, rather, I'd get rid of the counters. I don't think it's worth the code. And I'd rather not worry about the performance impact of the atomic increment either. For such a common increment we might want to look at per-cpu counters (which can be hacked in go with `sync.Pool` I think).
</blockquote></details>

I removed them.
___
*[pkg/util/hlc/hlc.go, line 122 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x8DZGEA1vpbOuTJMq:-M4D-3RZBUOns8r95vO2:b-hd66n) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/util/hlc/hlc.go#L122)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: s/Hlc/HLC everywhere
</blockquote></details>

changed everywhere.
___
*[pkg/util/timeutil/now_unix.go, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x5SUo0IK-MMBEtJtL:-M4D-JNm5AFbIqXhg-an:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/util/timeutil/now_unix.go#L28)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

just one `errors` pkg pls. The right one to use is `cockroachdb/errors`
</blockquote></details>

fixed
___
*[pkg/util/timeutil/now_unix.go, line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x26lPEWlLnkqIgiAV:-M4D-FMM1UYI3Gt0lXdA:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/util/timeutil/now_unix.go#L33)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

as the linter says, make it `clockDeviceID`
</blockquote></details>

done
___
*[pkg/util/timeutil/now_unix.go, line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x2IM_7jRfLVb-Lvl7:-M4D-GJP9M6HRjYS-UJM:bj5kf1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/util/timeutil/now_unix.go#L47)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Name? Path...
</blockquote></details>

added Path
___
*[pkg/util/timeutil/now_unix.go, line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x9Zbc7biYFjagOMJb:-M4D-MgL9HxbQiqE0Y5N:b-wkkfch) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/util/timeutil/now_unix.go#L47)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't like how this clock device works because it changes global state. And we really don't like globals. They always cause troubles - for example you can't run two servers in a process easily any more.
Everything should be scoped to a server. I think the clock device should be used to initialize an hlc clock - either make `hlc.NewClock()` take the device directly, or pass in a closure that uses the respective device as the existing `physicalClock` argument.

Also, I'd try to keep the new per-OS code out of `now.go`, in favor of keeping it in the hlc package. In fact, it seems to me like now.go can be un-split; there no longer seems to be a need for `now_windows.go` according to the comments in it. I think this would be a great time to see if we can clean that up.
</blockquote></details>

I changed to a closure.
___
*[pkg/util/timeutil/now_unix.go, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x6MxRBTdy1WTQvC5A:-M4D-L7_0MOlQFQmTLSN:b-yb1zw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/util/timeutil/now_unix.go#L54)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`clockDeviceFD`
</blockquote></details>

also capitalized
___
*[pkg/util/timeutil/now_unix.go, line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x2N8fCmIOaAkaa-s4:-M4D-HGv8Yq9YWhISvOJ:b-1i2b71) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/util/timeutil/now_unix.go#L55)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`clockID`
</blockquote></details>

capitalized
___
*[pkg/util/timeutil/now_unix.go, line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x55Pq9haDT221-5SY:-M4D-IRy42vHU4pT6ss_:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/util/timeutil/now_unix.go#L55)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please include a link explaining this magic here too
</blockquote></details>

done
___
*[pkg/util/timeutil/now_unix.go, line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3x6BruAG1BkCTSk8ey:-M4D-K8a8tRXWI3_pJzJ:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/util/timeutil/now_unix.go#L68)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`errors.Errorf()` pls. And don't capitalize the message.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> Reviewed 1 of 10 files at r1, 7 of 11 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @darinpp, and @knz)

___
*[build/builder/mkrelease.sh, line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3xCDQZ3Z31rFXgZGQP:-M4Dk6eY5GoORbgN_I4l:bvgbg00) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/build/builder/mkrelease.sh#L42)):*
<details><summary><i>Previously, darinpp wrote…</i></summary><blockquote>

It seems like that with the library versions that we are using it has to be explicit
</blockquote></details>

To reply to andrei's comment, `clock_gettime` was not part of the standard C library prior to glibc 2.17, and that's what our release builds use.

@darinpp I still think that Andrei's request is valid: please add a comment to explain. Your future self will thank you.
___
*[pkg/cli/cliflags/flags.go, line 680 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3xCa4w1mOcpRD_yfpy:-M4Dkmdv5d2_6Ph30WpR:b-9are1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/cli/cliflags/flags.go#L680)):*
<details><summary><i>Previously, darinpp wrote…</i></summary><blockquote>

Also mentioned that it is only for Linux. Also fixed the `+build` to be `linux` vs `!linux`. The validation is done already in `clock_gettime`. If you are using an incorrect device or you don't have such driver - it gives you back an error. 
</blockquote></details>

I'm ok with this behavior - Andrei if I am running a non-linux system and I happens to have the device available, I don't want my software to tell me "you're not using linux, go away". Dependency checks should be fine grained.
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4Dm9SZAJni8-n_Nn_F:-M4Dm9SZAJni8-n_Nn_G:bf0iauf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L28)):*
> ```Go
> )
> 
> // ClockDevice contains the handle of the clock device as well as the
> ```

I was trying the code out on my computer (which does not have a PTP device) and I got disappointed by the error message:

```
$ cockroach start --clock-device /dev/ptp0
ERROR: creating clock device: file does not exist: /dev/ptp0
```

The problem here is that the error messages gives me the impression that CockroachDB is the program creating /dev/ptp0 and the device in my OS.

I think the wording `creating clock device` in the error is poor.

This inconvenience then brought me to this file and now *as a programmer* I am confused: this struct `ClockDevice` *is not a device* - it is merely a handle for some abstraction provided by the operating system.

Consider: unix does not give you files, it gives you *file descriptors*. It does not give you IP connections, it gives you *sockets*. It does not give you memory, it gives you *pointers*.

You need a word here and in the error message that better explains what is going on here.

What about: `ClockSource` for the name of the struct:
- the fields *inside* the struct can remain unchanged: `ClockSource` abstracts a device from the OS into a source inside CockroachDB
- the command-line flag `--clock-device` can remain unchanged
- the contructour would be `MakeClockSource` 
- the error message would be `instantiating clock source`

what do you think?
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 48 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4Dmusp9XPIHgzMiWKu:-M4Dmusp9XPIHgzMiWKv:bqlgo37) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L48)):*
> ```Go
> 		log.Infof(
> 			context.TODO(),
> 			"Opened clock device %s with fd %x, mod_fd %x\n",
> ```

nit: log messages start with lowercase


<!-- Sent from Reviewable.io -->

andreimatei reviewed (approved):
LGTM with nits.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-:-M4FQG6E2E-vQBywifSM:b3590pt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @darinpp, and @knz)

___
*[build/builder/mkrelease.sh, line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3xCDQZ3Z31rFXgZGQP:-M4F7yDdBVelOMEsD3dd:b-6m0k5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/build/builder/mkrelease.sh#L42)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

To reply to andrei's comment, `clock_gettime` was not part of the standard C library prior to glibc 2.17, and that's what our release builds use.

@darinpp I still think that Andrei's request is valid: please add a comment to explain. Your future self will thank you.
</blockquote></details>

If that's true (and I guess it must be, cause a simple `make build` on Linux works without it, right?), consider adding a comment saying that we won't need to link this realtime library when we move to a newer libc.
___
*[pkg/cli/cliflags/flags.go, line 680 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3xCa4w1mOcpRD_yfpy:-M4F8iw30B4Tt02GcrH4:b-w96549) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/pkg/cli/cliflags/flags.go#L680)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I'm ok with this behavior - Andrei if I am running a non-linux system and I happens to have the device available, I don't want my software to tell me "you're not using linux, go away". Dependency checks should be fine grained.
</blockquote></details>

I'm pretty sure the API we're using is Linux specific - it comes from here: http://linuxptp.sourceforge.net/
I haven't found any references to BSD supporting it.
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4F9BXF3sActf94WGiD:-M4F9BXF3sActf94WGiE:b-2l22a8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L2)):*
> ```Go
> // Copyright 2017 The Cockroach Authors.
> //
> ```

s/linix/linux in file name
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4FD9Xl7N6AkC5AXkDy:-M4FD9Xl7N6AkC5AXkDz:b-oxv4dm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L31)):*
> ```Go
> // clock id.
> type ClockDevice struct {
> 	clockDevice   *os.File
> ```

is anybody using this field? We never close the file, do we? If nobody's using it, let's get rid of it.
And then we can just `type ClockDevice uintptr`.
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 44 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4F9xPQ3XThkkDFijHE:-M4F9xPQ3XThkkDFijHF:biheqqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L44)):*
> ```Go
> 		// For clarification of how the clock id is computed:
> 		// https://lore.kernel.org/patchwork/patch/868609/
> 		// https://github.com/torvalds/linux/blob/master/tools/testing/selftests/ptp/testptp.c#L87
> ```

links to a branch decay as the branch moves forward. Please put a a commit hash instead of `master`. You can easily get the url with the current one in Github by pressing `y` when you have the file open.
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4F9S4k5bx91ZzSUm0K:-M4F9S4k5bx91ZzSUm0L:b-zb59gj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L47)):*
> ```Go
> 		clockID := (^clockDeviceFD << 3) | 3
> 		log.Infof(
> 			context.TODO(),
> ```

no `context.TODO()` in new functions please. Take a `ctx`.
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4FB-qK0u-aJSjt368Y:-M4FB-qK0u-aJSjt368Z:bazmkr) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L50)):*
> ```Go
> 			"Opened clock device %s with fd %x, mod_fd %x\n",
> 			clockDevicePath,
> 			clockDeviceFD,
> ```

nit: I know that `File.Fd()` returns a `uintptr`, but I think it's weird to print it as hex. I believe I've only seen file descriptors printed as decimal - for example, in /proc/<pid>/fd.
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4FC45H8CYOuZpySPGU:-M4FC45H8CYOuZpySPGV:b-wd3hjs) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L50)):*
> ```Go
> 			"Opened clock device %s with fd %x, mod_fd %x\n",
> 			clockDevicePath,
> 			clockDeviceFD,
> ```

nit: consider getting rid of the two last arguments. They seem too in the weeds for an `Infof()`, and generally not very useful even at higher verbosity in my opinion.
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4FCHs5C2QcFE8Zgb0q:-M4FCHs5C2QcFE8Zgb0r:b-l33rsm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L56)):*
> ```Go
> 		_, err := C.clock_gettime(C.clockid_t(clockID), &ts)
> 		if err != nil {
> 			return result, errors.Wrap(err, "UseClockDevice: error calling clock_gettime")
> ```

say something about "invalid clock device"? That's why we're doing this reading here, isn't it?
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 60 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4FA_qf7YjPgII9aPv7:-M4FA_qf7YjPgII9aPv8:bat2cr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L60)):*
> ```Go
> 		result.clockDeviceID = clockID
> 	} else {
> 		return result, errors.Errorf("can't open %s, err %s\n", clockDevicePath, err)
> ```

invert the `if err == nil` condition, move this early return up, and unindent the main code path
___
*[pkg/util/hlc/hlc_clock_device_stub.go, line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4FDfoM0QqiPL920ZEe:-M4FDfoM0QqiPL920ZEf:b-smqvka) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_stub.go#L24)):*
> ```Go
> // clock id.
> type ClockDevice struct {
> 	clockDevice   *os.File
> ```

anybody using these fields?


<!-- Sent from Reviewable.io -->

darinpp reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @darinpp, and @knz)

___
*[build/builder/mkrelease.sh, line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M3xCDQZ3Z31rFXgZGQP:-M4G6ifGEqRWYee8MSDv:blswryk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3c14a21d5b3eac969051740be7a00325662a21/build/builder/mkrelease.sh#L42)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

If that's true (and I guess it must be, cause a simple `make build` on Linux works without it, right?), consider adding a comment saying that we won't need to link this realtime library when we move to a newer libc.
</blockquote></details>

I added a comment that we need `-lrt` this until we upgrade.
___
*[pkg/util/hlc/hlc_clock_device_stub.go, line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4FDfoM0QqiPL920ZEe:-M4G8sAHD5Cafz4SJF6F:b-2yscjt) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_stub.go#L24)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

anybody using these fields?
</blockquote></details>

Not explicitly. But this is needed to keep the device open and to prevent the GC from closing it.
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4F9BXF3sActf94WGiD:-M4G6y9YEe6fz47uvTtC:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L2)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

s/linix/linux in file name
</blockquote></details>

done
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4Dm9SZAJni8-n_Nn_F:-M4G70Jm1zpbcOzZZl85:bm2nd2x) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L28)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I was trying the code out on my computer (which does not have a PTP device) and I got disappointed by the error message:

```
$ cockroach start --clock-device /dev/ptp0
ERROR: creating clock device: file does not exist: /dev/ptp0
```

The problem here is that the error messages gives me the impression that CockroachDB is the program creating /dev/ptp0 and the device in my OS.

I think the wording `creating clock device` in the error is poor.

This inconvenience then brought me to this file and now *as a programmer* I am confused: this struct `ClockDevice` *is not a device* - it is merely a handle for some abstraction provided by the operating system.

Consider: unix does not give you files, it gives you *file descriptors*. It does not give you IP connections, it gives you *sockets*. It does not give you memory, it gives you *pointers*.

You need a word here and in the error message that better explains what is going on here.

What about: `ClockSource` for the name of the struct:
- the fields *inside* the struct can remain unchanged: `ClockSource` abstracts a device from the OS into a source inside CockroachDB
- the command-line flag `--clock-device` can remain unchanged
- the contructour would be `MakeClockSource` 
- the error message would be `instantiating clock source`

what do you think?
</blockquote></details>

changed the name to `ClockSource` and updated the error message
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4FD9Xl7N6AkC5AXkDy:-M4G75F3C-UeoYNIEpW8:b-kj3cav) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L31)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

is anybody using this field? We never close the file, do we? If nobody's using it, let's get rid of it.
And then we can just `type ClockDevice uintptr`.
</blockquote></details>

The field is needed to keep the file open. Otherwise GC will close the file.
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 44 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4F9xPQ3XThkkDFijHE:-M4G7ER_8W_Yan-IdQl9:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L44)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

links to a branch decay as the branch moves forward. Please put a a commit hash instead of `master`. You can easily get the url with the current one in Github by pressing `y` when you have the file open.
</blockquote></details>

done
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4F9S4k5bx91ZzSUm0K:-M4G7FYG45EDcopdF-VS:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L47)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

no `context.TODO()` in new functions please. Take a `ctx`.
</blockquote></details>

fixed
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 48 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4Dmusp9XPIHgzMiWKu:-M4G7GqX8_h9AUzbeJms:b-fwzznn) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L48)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: log messages start with lowercase
</blockquote></details>

changed to lowercase
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4FB-qK0u-aJSjt368Y:-M4G7JrsBD1NTDXKqjD7:b-cy0bl8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L50)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: I know that `File.Fd()` returns a `uintptr`, but I think it's weird to print it as hex. I believe I've only seen file descriptors printed as decimal - for example, in /proc/<pid>/fd.
</blockquote></details>

fixed it so the FD is printed as decimal.
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4FC45H8CYOuZpySPGU:-M4G7PqG2hXBmSuyzh1-:bgil2h5) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L50)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: consider getting rid of the two last arguments. They seem too in the weeds for an `Infof()`, and generally not very useful even at higher verbosity in my opinion.
</blockquote></details>

I disagree here. There was a situation in which the fd variable was shadowed and was defaulting to zero. The `clock_gettime` works fine in this case but uses the wrong source. You won't be able to determine that this is what happened unless you have this info. 
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4FCHs5C2QcFE8Zgb0q:-M4G7wv676vaVhCtd5sm:b2qw81y) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L56)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

say something about "invalid clock device"? That's why we're doing this reading here, isn't it?
</blockquote></details>

There could be a number of possible issues here. The device may not be invalid but there still may be a problem with getting the time. We are doing the call here to verify that getting the time works at least once. It is a lot less likely that it won't work after that first call succeeds. It will weed out the problems with missing kernel driver, incorrect device, permissions etc.
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 60 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4FA_qf7YjPgII9aPv7:-M4G8q4pETzGj4vjppEn:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L60)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

invert the `if err == nil` condition, move this early return up, and unindent the main code path
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

andreimatei reviewed (approved):
LGTM but please simplify the `!linux` `ClockDevice` struct.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-:-M4GHBNZ3EsH8sHhxBOC:b-e42v1k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @darinpp, and @knz)

___
*[pkg/util/hlc/hlc_clock_device_stub.go, line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4FDfoM0QqiPL920ZEe:-M4G9eE79S8XU376RJp9:bhaamvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_stub.go#L24)):*
<details><summary><i>Previously, darinpp wrote…</i></summary><blockquote>

Not explicitly. But this is needed to keep the device open and to prevent the GC from closing it.
</blockquote></details>

if we're never setting this field, what GC? Note that here I'm talking strictly about this `!linux` `ClockDevice`. I think it can be an empty struct.
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4FD9Xl7N6AkC5AXkDy:-M4GGvo6Do8SWWzLUZbp:b-c5ln1c) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L31)):*
<details><summary><i>Previously, darinpp wrote…</i></summary><blockquote>

The field is needed to keep the file open. Otherwise GC will close the file.
</blockquote></details>

Interesting. Please add a comment about it.
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4FC45H8CYOuZpySPGU:-M4GAZSG3oShf2eJ2ijK:b-duwvy6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L50)):*
<details><summary><i>Previously, darinpp wrote…</i></summary><blockquote>

I disagree here. There was a situation in which the fd variable was shadowed and was defaulting to zero. The `clock_gettime` works fine in this case but uses the wrong source. You won't be able to determine that this is what happened unless you have this info. 
</blockquote></details>

well it's not just any sort of programming errors that logs are meant to help you debug. Log are for dynamic conditions. 
Anyway, as you wish.
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4FCHs5C2QcFE8Zgb0q:-M4GAxKY3GABJNLm6A1O:btyi6w7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L56)):*
<details><summary><i>Previously, darinpp wrote…</i></summary><blockquote>

There could be a number of possible issues here. The device may not be invalid but there still may be a problem with getting the time. We are doing the call here to verify that getting the time works at least once. It is a lot less likely that it won't work after that first call succeeds. It will weed out the problems with missing kernel driver, incorrect device, permissions etc.
</blockquote></details>

If the device is "valid",  why would the call fail? A missing driver means an "invalid" device in my book. Permission errors are checked by the open call above, I guess. But even if they weren't, I'd still find an error talking about an invalid clock device and its path more useful than something telling me about `clock_gettime`.
But as you wish.
___
*[pkg/util/hlc/hlc_clock_device_linux.go, line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4GAHQA9ESDAyyrSbNC:-M4GAHQA9ESDAyyrSbND:bhw8s7k) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f6ad54572d111618c459a2e3d8eb44706193d30/pkg/util/hlc/hlc_clock_device_linux.go#L41)):*
> ```Go
> 	result.clockDevice, err = os.Open(clockDevicePath)
> 	if err != nil {
> 		return result, errors.Errorf("can't open %s, err %s\n", clockDevicePath, err)
> ```

I think you want `errors.Wrapf()`. And no `\n` pls.


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> also with the same nits as andrei

> Reviewed 1 of 11 files at r2, 4 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-:-M4GKn1S2-Gncj8kebi4:b-3mjiwu)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @darinpp, and @knz)

___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4FD9Xl7N6AkC5AXkDy:-M4GKOUU5VrRrPw538Ua:bmyza68) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L31)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Interesting. Please add a comment about it.
</blockquote></details>

I concur a comment would be useful here.
___
*[pkg/util/hlc/hlc_clock_device_linux.go, line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4GAHQA9ESDAyyrSbNC:-M4GKg5LALvTvb7vKGuk:bfxxbtd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f6ad54572d111618c459a2e3d8eb44706193d30/pkg/util/hlc/hlc_clock_device_linux.go#L41)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think you want `errors.Wrapf()`. And no `\n` pls.
</blockquote></details>

that's `errors.Wrapf(err, "cannot open %s", clockDevicePath)`


<!-- Sent from Reviewable.io -->

darinpp reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46942)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @darinpp, and @knz)

___
*[pkg/util/hlc/hlc_clock_device_stub.go, line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4FDfoM0QqiPL920ZEe:-M4GLWrXDkf2O-cl7cOx:ba8kcce) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_stub.go#L24)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

if we're never setting this field, what GC? Note that here I'm talking strictly about this `!linux` `ClockDevice`. I think it can be an empty struct.
</blockquote></details>

Ah. Got it. Removed both fields in the `!linux` case.
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4FD9Xl7N6AkC5AXkDy:-M4GPGAy8ZXI41of6N9_:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L31)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I concur a comment would be useful here.
</blockquote></details>

done
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4FC45H8CYOuZpySPGU:-M4GNmAf-RKbol44h4A7:b-r1g23c) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L50)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

well it's not just any sort of programming errors that logs are meant to help you debug. Log are for dynamic conditions. 
Anyway, as you wish.
</blockquote></details>

It isn't worth adding a separate `VEventf` just for these two when they are printed only once and the message makes sense to be there w/o them already. 
___
*[pkg/util/hlc/hlc_clock_device_linix.go, line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4FCHs5C2QcFE8Zgb0q:-M4GLbpt2bsv5O3UcPVS:b9wm0gk) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b7ab5d36ae124a92b8f1837eea6d6969a26b934/pkg/util/hlc/hlc_clock_device_linix.go#L56)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

If the device is "valid",  why would the call fail? A missing driver means an "invalid" device in my book. Permission errors are checked by the open call above, I guess. But even if they weren't, I'd still find an error talking about an invalid clock device and its path more useful than something telling me about `clock_gettime`.
But as you wish.
</blockquote></details>

My point is that I shouldn't make the determination that the device is invalid purely on the basis of it returning an error. The possible errors are not documented and may change over time. I'd rather log what the code tried doing - `calling clock_gettime` and what was the error. 
___
*[pkg/util/hlc/hlc_clock_device_linux.go, line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46942#-M4GAHQA9ESDAyyrSbNC:-M4GPLVr9mudeAgM8Bpt:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f6ad54572d111618c459a2e3d8eb44706193d30/pkg/util/hlc/hlc_clock_device_linux.go#L41)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

that's `errors.Wrapf(err, "cannot open %s", clockDevicePath)`
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

darinpp commented:
bors r+

darinpp commented:
bors r+
